packages/fxa-auth-server/scripts/cancel-subscriptions-to-plan/cancel-subscriptions-to-plan.ts

```diff
@@ -0,0 +1,261 @@
+    private excludePlanIds: string[],
+    private batchSize: number,
+    private outputFile: string,
+    private stripeHelper: StripeHelper,
+    private database: any,
+    public dryRun: boolean,
+    rateLimit: number
+  ) {
+    this.stripe = this.stripeHelper.stripe;
+
+    const config = Container.get<ConfigType>(AppConfig);
+    this.config = config;
+
+    const firestore = Container.get<Firestore>(AuthFirestore);
+    this.firestore = firestore;
+
+    this.stripeQueue = new PQueue({
+      intervalCap: rateLimit,
+      interval: 1000, // Stripe measures its rate limit per second
+    });
+  }
+
+  /**
+   * Cancel all customer subscriptions in batches
+   */
+  async run() {
+    let startAfter: string | null = null;
+    let hasMore = true;
+
+    while (hasMore) {
+      const subscriptions = await this.fetchSubsBatch(startAfter);
+
+      startAfter = subscriptions.at(-1)?.id as string;
+      if (!startAfter) hasMore = false;
+
+      await Promise.all(
+        subscriptions.map((sub) => this.processSubscription(sub))
+      );
+    }
+  }
+
+  /**
+   * Fetches subscriptions from Firestore paginated by batchSize
+   * @param startAfter ID of the last element of the previous batch for pagination
+   * @returns A list of subscriptions from firestore
+   */
+  async fetchSubsBatch(startAfter: string | null) {
+    const collectionPrefix = `${this.config.authFirestore.prefix}stripe-`;
+    const subscriptionCollection = `${collectionPrefix}subscriptions`;
+
+    const subscriptionSnap = await this.firestore
+      .collectionGroup(subscriptionCollection)
+      .where('plan.id', '==', this.priceId)
+      .orderBy('id')
+      .startAfter(startAfter)
+      .limit(this.batchSize)
+      .get();
+
+    const subscriptions = subscriptionSnap.docs.map(
+      (doc) => doc.data() as FirestoreSubscription
+    );
+
+    return subscriptions;
+  }
+
+  /**
+   * Attempts to cancel a firestore subscription
+   * @param firestoreSubscription The subscription to cancel
+   */
+  async processSubscription(firestoreSubscription: FirestoreSubscription) {
+    const { id: subscriptionId, customer: customerId } = firestoreSubscription;
+
+    try {
+      const customer = await this.fetchCustomer(customerId);
+      if (!customer?.subscriptions?.data) {
+        console.error(`Customer not found: ${customerId}`);
+        return;
+      }
+
+      const account = await this.database.account(customer.metadata.userid);
+      if (!account) {
+        console.error(`Account not found: ${customer.metadata.userid}`);
+        return;
+      }
+
+      const isExcluded = this.isCustomerExcluded(customer.subscriptions.data);
+
+      if (!this.dryRun && !isExcluded) {
+        await this.cancelSubscription(firestoreSubscription);
+      }
+
+      const report = this.buildReport(customer, account, isExcluded);
+
+      await this.writeReport(report);
+
+      console.log(subscriptionId);
+    } catch (e) {
+      console.error(subscriptionId, e);
```

Comment:
If we are capturing the success to the report file, do we want to capture the errors to anything other than std err?

Reply:
Nope - since the subscription is likely unaffected, we mustn't include them in the CSV for marketing. I use stdout/stderr for all message parsing and analytics for the results of the script run. The stdout/stderr is captured by SRE, so if there are errors we'll be able to use the information dumped here. The cancellation of any particular subscription also shouldn't block the script from continuing.